File: java/tst/tarehart/rlbot/tuning/PredictedVsActualTest.java
package tarehart.rlbot.tuning;

import com.google.gson.Gson;
import tarehart.rlbot.math.TimeUtil;
import tarehart.rlbot.math.vector.Vector2;
import tarehart.rlbot.math.vector.Vector3;
import org.junit.Assert;
import org.junit.Test;
import tarehart.rlbot.math.BallSlice;
import tarehart.rlbot.physics.ArenaModel;
import tarehart.rlbot.physics.BallPath;

import java.io.InputStream;
import java.io.UnsupportedEncodingException;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Scanner;

public class PredictedVsActualTest {


    private static final double THRESHOLD = 4;
    private ArenaModel arenaModel = new ArenaModel();


    private BallPath readRecording(String filename) {
        InputStream in = getClass().getResourceAsStream("/ballrecordings/" + filename);
        Scanner s = new Scanner(in).useDelimiter("\\A");
        String content = s.hasNext() ? s.next() : "";

        Gson gson = new Gson();
        return gson.fromJson(content, BallPath.class);
    }

    @Test
    public void spinlessBackwall() throws UnsupportedEncodingException {
        testFile("spinless-backwall.json");
    }

    @Test
    public void spinlessGround() throws UnsupportedEncodingException {
        testFile("spinless-ground.json");
    }

    private void testFile(String filename) throws UnsupportedEncodingException {

        BallPath actualPath = readRecording(filename);
        actualPath = finesseActualPath(actualPath);
        BallPath predictedPath = makePrediction(actualPath);
        // (-73.29997, 65.447556, 4.5342107) after first time step

        List<BallSlice> actual = actualPath.getSlices();
        for (int i = 0; i < actual.size(); i++) {
            if (i < 20) {
                Vector3 velocity = actual.get(i).getVelocity();
                double speedNow = velocity.magnitude();

                Vector3 velocityAfter = actual.get(i+1).getVelocity();
                double speedNext = velocityAfter.magnitude();

                double drag = speedNext / speedNow;
                double dragPerSpeed = drag / speedNow;
                //System.out.println(String.format("Velocity: %s Speed: %s Drag: %s DragPerSpeed: %s", velocity, speedNow, drag, dragPerSpeed));
            }
        }

        List<BallSlice> predictedSlices = predictedPath.getSlices();

        List<BallSlice> actualTrimmed = new ArrayList<>(predictedSlices.size());

        for (int i = 0; i < predictedSlices.size() - 1; i++) {
            BallSlice actualSlice = actualPath.getMotionAt(predictedSlices.get(i).getTime()).get();
            actualTrimmed.add(actualSlice);
            System.out.println(String.format("A: %s\nP: %s\n", actualSlice, predictedSlices.get(i)));

        }
        actualTrimmed.add(actualPath.getEndpoint());

        for (int i = 0; i < predictedSlices.size(); i++) {

            Vector3 actualSlice = actualTrimmed.get(i).getSpace();
            Vector3 actualToPredicted = predictedSlices.get(i).getSpace().minus(actualSlice);
            double error = new Vector2(actualToPredicted.x, actualToPredicted.y).magnitude();
            if (error > THRESHOLD) {
                Duration duration = Duration.between(actualTrimmed.get(0).getTime(), actualTrimmed.get(i).getTime());
                double seconds = duration.toMillis() / 1000.0;
                Assert.fail(String.format("Diverged to %.2f after %.2f seconds!", error, seconds));
            }
        }
    }

    private BallPath finesseActualPath(BallPath actualPath) {
        Optional<BallSlice> newStart = actualPath.getMotionAt(actualPath.getStartPoint().getTime().plus(Duration.ofMillis(100)));
        BallPath finessed = new BallPath(newStart.get());
        List<BallSlice> slices = actualPath.getSlices();
        for (int i = 0; i < actualPath.getSlices().size(); i++) {
            if (slices.get(i).getTime().isAfter(newStart.get().getTime())) {
                finessed.addSlice(slices.get(i));
            }
        }
        return finessed;
    }

    private BallPath makePrediction(BallPath backWallActual) {
        Duration duration = Duration.between(backWallActual.getStartPoint().getTime(), backWallActual.getEndpoint().getTime());
        return arenaModel.simulateBall(backWallActual.getStartPoint(), duration);
    }

}

File: java/tst/tarehart/rlbot/tuning/PredictedVsActualTest.kt
package tarehart.rlbot.tuning

import com.google.gson.Gson
import org.junit.Assert
import org.junit.Test
import tarehart.rlbot.math.BallSlice
import tarehart.rlbot.math.vector.Vector2
import tarehart.rlbot.math.vector.Vector3
import tarehart.rlbot.physics.ArenaModel
import tarehart.rlbot.physics.BallPath
import tarehart.rlbot.time.Duration

import java.io.InputStream
import java.io.UnsupportedEncodingException
import java.util.ArrayList
import java.util.Optional
import java.util.Scanner

class PredictedVsActualTest {
    private val arenaModel = ArenaModel()


    private fun readRecording(filename: String): BallPath {
        val input = javaClass.getResourceAsStream("/ballrecordings/" + filename)
        val s = Scanner(input).useDelimiter("\\A")
        val content = if (s.hasNext()) s.next() else ""

        val gson = Gson()
        return gson.fromJson(content, BallPath::class.java)
    }


    // These files were deleted because the format became obsolete. We should record some new ones.

    //    @Test
    //    public void spinlessBackwall() throws UnsupportedEncodingException {
    //        testFile("spinless-backwall.json");
    //    }
    //
    //    @Test
    //    public void spinlessGround() throws UnsupportedEncodingException {
    //        testFile("spinless-ground.json");
    //    }

    @Throws(UnsupportedEncodingException::class)
    private fun testFile(filename: String) {

        var actualPath = readRecording(filename)
        actualPath = finesseActualPath(actualPath)
        val predictedPath = makePrediction(actualPath)
        // (-73.29997, 65.447556, 4.5342107) after first time step

        val actual = actualPath.slices
        for (i in actual.indices) {
            if (i < 20) {
                val velocity = actual[i].velocity
                val speedNow = velocity.magnitude()

                val velocityAfter = actual[i + 1].velocity
                val speedNext = velocityAfter.magnitude()

                val drag = speedNext / speedNow
                val dragPerSpeed = drag / speedNow
                //System.out.println(String.format("Velocity: %s Speed: %s Drag: %s DragPerSpeed: %s", velocity, speedNow, drag, dragPerSpeed));
            }
        }

        val predictedSlices = predictedPath.slices

        val actualTrimmed = ArrayList<BallSlice>(predictedSlices.size)

        for (i in 0 until predictedSlices.size - 1) {
            val actualSlice = actualPath.getMotionAt(predictedSlices[i].time)!!
            actualTrimmed.add(actualSlice)
            println(String.format("A: %s\nP: %s\n", actualSlice, predictedSlices[i]))

        }
        actualTrimmed.add(actualPath.endpoint)

        for (i in predictedSlices.indices) {

            val actualSlice = actualTrimmed[i].space
            val (x, y) = predictedSlices[i].space.minus(actualSlice)
            val error = Vector2(x, y).magnitude()
            if (error > THRESHOLD) {
                val duration = Duration.between(actualTrimmed[0].time, actualTrimmed[i].time)
                Assert.fail(String.format("Diverged to %.2f after %.2f seconds!", error, duration.seconds))
            }
        }
    }

    private fun finesseActualPath(actualPath: BallPath): BallPath {
        val newStart = actualPath.getMotionAt(actualPath.startPoint.time.plus(Duration.ofMillis(100)))
        val finessed = BallPath(newStart!!)
        val slices = actualPath.slices
        for (i in 0 until actualPath.slices.size) {
            if (slices[i].time.isAfter(newStart.time)) {
                finessed.addSlice(slices[i])
            }
        }
        return finessed
    }

    private fun makePrediction(backWallActual: BallPath): BallPath {
        val duration = Duration.between(backWallActual.startPoint.time, backWallActual.endpoint.time)
        return arenaModel.simulateBall(backWallActual.startPoint, duration)
    }

    companion object {


        private val THRESHOLD = 4.0
    }

}
